--- openssl-3.0.7/apps/fipsinstall.c    2023-06-26 14:39:03.349903094 +0000
+++ openssl-3.0.7.new/apps/fipsinstall.c        2023-06-26 14:45:11.366113292 +0000
@@ -311,7 +311,7 @@
     EVP_MAC *mac = NULL;
     CONF *conf = NULL;
 
-    BIO_printf(bio_err, "This command is not enabled in the Red Hat Enterprise Linux OpenSSL build, please consult Red Hat documentation to learn how to enable FIPS mode\n");
+    BIO_printf(bio_err, "This command is not enabled in OpenSSL build, please consult documentation to learn how to enable FIPS mode\n");
     return 1;
 
     if ((opts = sk_OPENSSL_STRING_new_null()) == NULL)
